pennylane/templates/subroutines/qmc.py

```diff
@@ -326,32 +325,47 @@ def circuit():
         >>> (1 - np.cos(np.pi * phase_estimated)) / 2
         0.4327096457464369
     """
-    if isinstance(probs, np.ndarray) and probs.ndim != 1:
-        raise ValueError("The probability distribution must be specified as a flat array")
-
-    dim_p = len(probs)
-    num_target_wires = float(np.log2(2 * dim_p))
-
-    if not num_target_wires.is_integer():
-        raise ValueError("The probability distribution must have a length that is a power of two")
-
-    num_target_wires = int(num_target_wires)
-
-    target_wires = Wires(target_wires)
-    estimation_wires = Wires(estimation_wires)
-
-    if num_target_wires != len(target_wires):
-        raise ValueError(
-            f"The probability distribution of dimension {dim_p} requires"
-            f" {num_target_wires} target wires"
-        )
-
-    A = probs_to_unitary(probs)
-    R = func_to_unitary(func, dim_p)
-    Q = make_Q(A, R)
-
-    qml.QubitUnitary(A, wires=target_wires[:-1])
-    qml.QubitUnitary(R, wires=target_wires)
-    qml.templates.QuantumPhaseEstimation(
-        Q, target_wires=target_wires, estimation_wires=estimation_wires
-    )
+    num_params = 3
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, probs, func, target_wires, estimation_wires, do_queue=True):
+        if isinstance(probs, np.ndarray) and probs.ndim != 1:
+            raise ValueError("The probability distribution must be specified as a flat array")
+
+        dim_p = len(probs)
+        num_target_wires = float(np.log2(2 * dim_p))
```

Comment:
@trbromley this template looks like it might break if used with the torch/tf interfaces, and won't be differentiable using the autograd interfaces.

Currently, most other templates are written to be differentiable (even Mottonen!). Could you re-write this to ensure differentiability is preserved?

Reply:
:thinking: I think this may be highly non-trivial, we'd need the following:
- `qml.math.outer`
- `qml.math.eye`
- `qml.math.kron`

I'd also have the following questions:
- Does `qml.QubitUnitary` support differentiation? Maybe this could be supported in backprop mode :thinking: 
- I like the "differentiable everywhere" philosophy, but I'm not sure why one would want to do this for QMC. You are calculating `sum_i p(i) f(i)`, so I guess we'd be looking at calculating `sum_i dp(i)/dx f(i)`?
- How would it work with the callable function `func`? Would that need to be interface compatible, e.g., with a forward and backward method?

So overall, since this is a template/operation that effectively loads fixed unitary matrices based upon an input probs and function, I'm not sure if it fits since it's not a variational circuit (at least how I usually think about them)

Reply:
> qml.math.outer, qml.math.eye,  qml.math.kron

`qml.math.eye` - should be fine to just use `np.eye`, and then cast it (since it will be a 'constant' array).

The other two could probably be supported if we added `qml.math.tensordot` 🤔 

> Does qml.QubitUnitary support differentiation? Maybe this could be supported in backprop mode 🤔

Not yet, but this is on our todo list (it has come up in user bugs). The reason that it doesn't support differentiation is the same; the `_matrix()` method calls `U.conj().T` rather than `qml.math.transpose(qml.math.conj(U))`:

```python
if not np.allclose(U @ U.conj().T, np.identity(U.shape[0])):
            raise ValueError("Operator must be unitary.")
```

> I like the "differentiable everywhere" philosophy, but I'm not sure why one would want to do this for QMC. You are calculating sum_i p(i) f(i), so I guess we'd be looking at calculating sum_i dp(i)/dx f(i)?

Oh, on your other points... yes, this makes sense. Perhaps 'differentiable everywhere' is out of scope for QMC.

I saw vanilla numpy, and my brain originally thought 'this breaks differentiability!'

On that end, though, I wonder if there is a way to raise a reasonable error message _if_ someone tries to differentiate it 🤔 I don't think there is.

Reply:
I also agree with Tom...but one could add a warning to the docstring?

Reply:
Nice, I mentioned in the current note: 68cfb6f0d3ce9de21a0ed5eba80519a2c3c4ae7e